src/eos-updater-apply.c

```diff
@@ -73,6 +74,431 @@ get_test_osname (void)
   return g_getenv ("EOS_UPDATER_TEST_UPDATER_OSTREE_OSNAME");
 }
 
+typedef struct
+{
+  char *repo;
+  char *app_identifier;
+  FlatpakRefKind ref_kind;
+} FlatpakToInstall;
+
+static FlatpakToInstall *
+flatpak_to_install_new_steal (char           *repo,
+                              char           *app_identifier,
+                              FlatpakRefKind kind)
+{
+  FlatpakToInstall *to_install = g_new0 (FlatpakToInstall, 1);
+
+  to_install->repo = repo;
+  to_install->app_identifier = app_identifier;
+
+  return to_install;
+}
+
+static FlatpakToInstall *
+flatpak_to_install_copy (FlatpakToInstall *to_install)
+{
+  return flatpak_to_install_new_steal (g_strdup (to_install->repo),
+                                       g_strdup (to_install->app_identifier),
+                                       to_install->ref_kind);
+}
+
+static void
+flatpak_to_install_free (gpointer data)
+{
+  FlatpakToInstall *to_install = data;
+
+  g_free (to_install->repo);
+  g_free (to_install->app_identifier);
+  g_free (to_install);
+}
+
+static void
+malformed_flatpak_autoinstall_file_error (GError **error,
+                                          const gchar *entry,
+                                          const gchar *path)
+{
+  g_set_error (error,
+               EOS_UPDATER_ERROR,
+               EOS_UPDATER_ERROR_MALFORMED_FLATPAK_ENTRIES,
+               "Malformed entry %s in flatpak auto-install file %s, "
+               "please consult your vendor",
+               entry,
+               path);
+}
+
+static gchar *
+get_temporary_directory_to_checkout_in (GError **error)
+{
+  g_autofree gchar *temp_dir = g_dir_make_tmp ("ostree-checkout-XXXXXX", error);
+
+  if (!temp_dir)
+    return NULL;
+
+  return g_build_filename (temp_dir, "checkout", NULL);
+}
+
+static gboolean
+parse_flatpak_ref_kind (const gchar *string, FlatpakRefKind *kind)
```

Comment:
Ditto, would be nice to use a more standard method for this and the following; at least they should be shared with similar code in the flatpak installer